### PR TITLE
Lb2 default policy tests

### PR DESCRIPTION
**Pull request by wprzytula**
Latency awareness tests are still to be done.

### CODE REVIEW

File: scylla/src/transport/load_balancing/default.rs [hunk 1]
```diff
@@ -1400,3 +1400,167 @@ mod latency_awareness {
...
+            while let Some(next_got) = got.next() {
+                if !curr_group.remove(&next_got) {
+                    panic!("Received unexpected node {}", next_got);
+                }
+
+                if curr_group.is_empty() {
+                    match (got.peek(), expected_groups.next()) {
+                        (None, None) => return, // Test succeeded
+                        (None, Some(group)) => panic!(
+                            "Not enough groups present in the plan; first non matched group: {:?}",
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs:1487]: I think it should be: `Not enough nodes present in the plan`

File: scylla/src/transport/load_balancing/default.rs [hunk 2]
```diff
@@ -1400,3 +1400,167 @@ mod latency_awareness {
...
+                            group
+                        ),
+                        (Some(node), None) => panic!("Unexpected node {} in the plan", node),
+                        (Some(_), Some(group)) => curr_group = group,
+                    }
+                }
+            }
+            panic!(
+                "Not enough groups present in the plan; first non matched group: {:?}",
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs:1496]: Also `s/groups/nodes`

File: scylla/src/transport/load_balancing/default.rs [hunk 3]
```diff
@@ -1400,3 +1400,167 @@ mod latency_awareness {
...
+            );
+        }
+
+        #[test]
+        fn test_assert_proper_grouping_in_plan_good() {
+            let got = vec![1u16, 2, 3, 4, 5];
+            let expected_groups = [&[1u16] as &[u16], &[3, 2, 4], &[5]]
+                .into_iter()
+                .map(|nodes| nodes.into_iter().copied().collect::<HashSet<u16>>())
+                .collect::<Vec<_>>();
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs:1507]: The tests could use the `ExpectedGroupsBuilder`, but I guess it doesn't really matter.

File: scylla/src/transport/load_balancing/default.rs [hunk 4]
```diff
@@ -1400,3 +1400,167 @@ mod latency_awareness {
...
+        }
+
+        // based on locator mock cluster
+        pub(crate) async fn mock_cluster_data_for_token_aware_tests() -> ClusterData {
+            let metadata = mock_metadata_for_token_aware_tests();
+            ClusterData::new(metadata, &Default::default(), &HashMap::new(), &None, None).await
+        }
+
+        // creates ClusterData with info about 5 nodes living in 2 different datacenters
+        // ring field is empty
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs]: Is it true that the ring field is empty? Each node has a single token and I think `ClusterData` creates the ring using these tokens. The name `with_no_token_info` suggests that the nodes don't have any tokens.

File: scylla/src/transport/load_balancing/default.rs [hunk 5]
```diff
@@ -1563,4 +1583,315 @@ mod tests {
...
+        fn default() -> Self {
+            Self {
+                preferred_datacenter: None,
+                is_token_aware: true,
+                permit_dc_failover: true,
+                latency_awareness: None,
+                pick_predicate: Box::new(Self::is_alive),
+            }
+        }
+    }
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs:1599]: Let's move the `Default` impl out of the `tests` module.

File: scylla/src/transport/load_balancing/default.rs [hunk 6]
```diff
@@ -1563,4 +1583,315 @@ mod tests {
...
+
+    pub(crate) const EMPTY_QUERY_INFO: QueryInfo = QueryInfo {
+        token: None,
+        keyspace: None,
+        is_confirmed_lwt: false,
+        consistency: Consistency::Quorum,
+        serial_consistency: Some(SerialConsistency::Serial),
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs:1606]: I think the default values are `LocalQuorum` and `LocalSerial`, although it probably doesn't matter here.

File: scylla/src/transport/load_balancing/default.rs [hunk 7]
```diff
@@ -1563,4 +1583,315 @@ mod tests {
...
+
+        for plan in plans {
+            println!(
+                "Got: {:?}, expected: {:?}",
+                &plan,
+                expected_groups.clone().collect::<Vec<_>>()
+            );
+            assert_proper_grouping_in_plan(plan.into_iter(), expected_groups.clone());
+        }
+    }
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs]: Why do the same thing twice?

File: scylla/src/transport/load_balancing/default.rs [hunk 8]
```diff
@@ -1563,4 +1583,315 @@ mod tests {
...
+                },
+                expected_groups: ExpectedGroupsBuilder::new()
+                    .group([A, B, C, G]) // local nodes
+                    .group([D, E, F]) // remote nodes
+                    .build(),
+            },
+            Test {
+                // Unknown preferred DC, failover permitted
+                policy: DefaultPolicy {
+                    preferred_datacenter: Some("au".to_owned()),
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs:1886]: Could this be set to `None`?

File: scylla/src/transport/load_balancing/default.rs [hunk 9]
```diff
@@ -1563,4 +1583,315 @@ mod tests {
...
+                    token: Some(Token { value: 160 }),
+                    keyspace: Some(KEYSPACE_NTS_RF_2),
+                    consistency: Consistency::Quorum,
+                    ..Default::default()
+                },
+                // going though the ring, we get order: F , A , C , D , G , B , E
+                //                                      us  eu  eu  us  eu  eu  us
+                //                                      r2  r1  r1  r1  r2  r1  r1
+                expected_groups: ExpectedGroupsBuilder::new().build(), // empty plan, because all nodes are remote and failover is forbidden
+            },
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs:1923]: Maybe it could also be tested what happens when the keyspace is not available.

**Review by cvybhu — COMMENTED** [review]
Looks good,
although it would easier to review if there were more comments/commit messages. For example why does pick() return an `Option` now? What has changed in the design?

File: scylla/src/transport/load_balancing/default.rs (hunk 1 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:1487]: It should indeed be as I wrote. In current DefaultPolicy implementation, nodes are grouped. If a node misses from a group, then another error message is emitted. In this case, however, the whole node group is missing, hence the message.

File: scylla/src/transport/load_balancing/default.rs (hunk 3 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:1507]: The tests were intended to test the function `assert_proper_grouping`, not the builder.

File: scylla/src/transport/load_balancing/default.rs (hunk 4 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs]: Good catch, I've reused the cluster from old LB tests. The problem is that in the old implementation an empty token ring was not a problem, whereas with DefaultPolicy taking advantage of locator, the ring cannot be empty of else we panic. I'll think about that.

File: scylla/src/transport/load_balancing/default.rs (hunk 5 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:1599]: I've put it there on purpose, as it may be tempting to use this default impl by the users, this way losing possibility to e.g. set the preferred datacenter. I thought that the users should always use the builder, thence being aware of the available configuration options.

File: scylla/src/transport/load_balancing/default.rs (hunk 6 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:1606]: Local consistencies forbid the datacenter failover, which we most often want to happen in tests.

File: scylla/src/transport/load_balancing/default.rs (hunk 7 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs]: The `println` is a residue. I'll remove it.

File: scylla/src/transport/load_balancing/default.rs (hunk 8 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:1886]: Of course, but the very goal of this test is to check the DP's behaviour when `preferred_datacenter` is set to an unknown DC.

File: scylla/src/transport/load_balancing/default.rs (hunk 2 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:1496]: See my previous response.

File: scylla/src/transport/load_balancing/default.rs (hunk 9 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:1923]: It is already tested in one test, in line 1815 (look for `// No keyspace implies no token awareness`)

File: scylla/src/transport/load_balancing/default.rs (hunk 4 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs]: > The name with_no_token_info suggests that the nodes don't have any tokens.

The name was to suggest that the `QueryInfo` provided to `DefaultPolicy` contains no Token info, that is, the query was not prepared. I'll adjust the name.

File: scylla/src/transport/load_balancing/default.rs (hunk 1 above)
Comment (cvybhu) [reply scylla/src/transport/load_balancing/default.rs:1487]: Well a load balancing plan is a list of nodes, not groups. We look for the next node in the plan and if its not there we fail. If I saw this error message I would assume that I passed to little `expected_groups` to the function.

File: scylla/src/transport/load_balancing/default.rs (hunk 5 above)
Comment (cvybhu) [reply scylla/src/transport/load_balancing/default.rs:1599]: Hmm at first I thought it does't matter where the impl is defined, it doesn't change the visiblity, but because of `#[cfg(test)]` it's only visible to tests. Makes sense :+1:

File: scylla/src/transport/load_balancing/default.rs (hunk 6 above)
Comment (cvybhu) [reply scylla/src/transport/load_balancing/default.rs:1606]: Then there should be tests for it

File: scylla/src/transport/load_balancing/default.rs (hunk 9 above)
Comment (cvybhu) [reply scylla/src/transport/load_balancing/default.rs:1923]: Right, missed it :+1:

File: scylla/src/transport/load_balancing/default.rs (hunk 8 above)
Comment (cvybhu) [reply scylla/src/transport/load_balancing/default.rs:1886]: I think it would be good to add a test where `preferred_datacenter` is None, I don't see one.

**Comment by cvybhu** [comment]
All these tests use `NetworkTopologyStrategy`, we should also test `SimpleStrategy`

File: scylla/src/transport/load_balancing/default.rs (hunk 6 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:1606]: Test `// Keyspace with RF=2 with DC failover forbidden by local Consistency` checks it.

**Comment by wprzytula** [comment]
> All these tests use `NetworkTopologyStrategy`, we should also test `SimpleStrategy`

SimpleStrategy is going to be deleted from Scylla, as Yaniv noticed. Should we really invest time in testing it thoroughly? It is already tested in the locator module, so if NTS works well in DefaultPolicy, SimpleStrategy will will do as well.

File: scylla/src/transport/load_balancing/default.rs (hunk 6 above)
Comment (cvybhu) [reply scylla/src/transport/load_balancing/default.rs:1606]: Right, missed it too x.x

File: scylla/src/transport/load_balancing/default.rs (hunk 1 above)
Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:1487]: A load balancing plan of a DefaultPolicy is a flat list of groups of nodes, or concatenation of groups beings sets of nodes. If you saw this error message, you should think this way: tests never assume that the expected input is bad. In case of inequality, it is the actual input that is to be blamed for being invalid. Thus: the message concerns the _actual plan_, not the _expected groups_.

**Comment by cvybhu** [comment]
I think it will be a loong time until SimpleStrategy is removed, it might be in the plans, but actually phasing it out will require great effort.

Locator takes care of finding the replicas, the load balancing policy has more responsibilities that the locator tests don't test. For example, how should the SimpleStrategy behave with a preferred_datacenter? Does it ignore disabled nodes?

File: scylla/src/transport/load_balancing/default.rs (hunk 1 above)
Comment (cvybhu) [reply scylla/src/transport/load_balancing/default.rs:1487]: Well it's `got: impl Iterator<Item = u16>,`, an iterator over individual nodes, not groups. I still think the message is misleading, but I don't care enough to continue arguing about it.

**Comment by havaker** [comment]
> I think it will be a loong time until SimpleStrategy is removed, it might be in the plans, but actually phasing it out will require great effort.

I agree we should test it.

> For example, how should the SimpleStrategy behave with a preferred_datacenter?

Not returning remote nodes would be the most reasonable behavior I think. I wolud also log some warning when a combination of `Some(preferred_dc)` and a SimpleStrategy is encoutered, as it may lead to empty query plans.

**Comment by wprzytula** [comment]
> Looks good, although it would easier to review if there were more comments/commit messages. For example why does pick() return an `Option` now? What has changed in the design?

The change is caused by our conclusion that we should never return a node from a remote DC if a preferred DC is set to Some. In the case that we provide an unknown preferred DC, we should return an empty plan, which requires that pick() return Option.

**Comment by wprzytula** [comment]
> > I think it will be a loong time until SimpleStrategy is removed, it might be in the plans, but actually phasing it out will require great effort.
> 
> I agree we should test it.
> 
> > For example, how should the SimpleStrategy behave with a preferred_datacenter?
> 
> Not returning remote nodes would be the most reasonable behavior I think. I wolud also log some warning when a combination of `Some(preferred_dc)` and a SimpleStrategy is encoutered, as it may lead to empty query plans.

Please remember then to include the recommendation not to combine preferred DC with SimpleStrategy in the docs.

**Comment by wprzytula** [comment]
> > I think it will be a loong time until SimpleStrategy is removed, it might be in the plans, but actually phasing it out will require great effort.
> 
> I agree we should test it.
> 
> > For example, how should the SimpleStrategy behave with a preferred_datacenter?
> 
> Not returning remote nodes would be the most reasonable behavior I think. I wolud also log some warning when a combination of `Some(preferred_dc)` and a SimpleStrategy is encoutered, as it may lead to empty query plans.

Is a combination of `SimpleStrategy` and `Some(preferred_dc)` enought to warn, or should we require also `permit_dc_failover` set to `false`?

**Comment by havaker** [comment]
> Is a combination of SimpleStrategy and Some(preferred_dc) enought to warn, or should we require also permit_dc_failover set to false?

I think the latter option is better.

**Review by havaker — APPROVED** [review]
